rfcs/2023-04-18-bundle-check.md

```diff
@@ -0,0 +1,124 @@
+- This github action will pass or fail based on the limits defined in the config.
+- To get a diff of master we will need to allow Bundlewatch to add a build-status check to the PR (requires github admin approval). The status check will show the diff between the current bundle size and the size of the bundle on master. This will help us to visualize the impact of our changes on the bundle size.
+- The build-status check will show the bundle size like this:
+<img src="./images/bundle-size/build-status.png" width="800" />
+
+- When you click on the `Details` link, it will take you to a page like this:
+<img src="./images/bundle-size/bundle-details.png" width="800" />
+
+- This does not post any comments on the PR to show the diff or the bundle size, we will have to write some code to extract bundle size from the Github Action, curate a custom comment and post it on the PR.
+
+### Tree Shaking Check
+We will achieve the following with this:
+
+3. Ensure our tree-shaking works as expected by adding a check to ensure that unused components are not included in the bundle.
+
+- We will use [size-limit](https://github.com/ai/size-limit) for this
+- size-limit is a tool that helps us to keep an eye on our bundle size. It works by comparing the size of our bundles against a predefined limit. If the size of the bundle exceeds the limit, the build fails.
+- size-limit creates an empty project and imports the component that we want. It then creates a bundle and checks the size of the bundle. If the size of the bundle exceeds the limit, the build fails.
+- This will enable us to limit importing only a small set of components like `{ Button, Text }` and check the bundle size to ensure that tree-shaking is working as intended and the overall bundle size is limited to the imported components.
+- size-limit configuration looks like this:
+```json
+  "size-limit": [
+    {
+      "path": "./build/components/index.web.js",
+      "import": "{ Button, Text }",
+      "limit": "100 kB"
+    }
+  ]
+```
+- We are not using size-limit to do a total bundle size check or master diff because it doesn't support it as well as bundlewatch does. Bundlewatch has an infra set up that stores master bundle size and diffs it on every PR. We can't do that with size-limit.
+
+# Drawbacks/Constraints
+- Bundlewatch is a free service but it also seems unmaintained since Jan 2022.
+- If Bundlewatch service goes down, we will not be able to get any master diffs but we will still be able to get the total bundle size limit check.
+- Bundlewatch service is also open source: https://github.com/bundlewatch/service
+- We could spin up our own service to store master bundle size and diff it on every PR but that would be a lot of work and we would have to maintain it.
+- Alternate to using bundlewatch is setting up our own infra to store master bundle size and diff it on every PR. This would also be a lot of work and we would have to maintain it.
```

Comment:
I think we just need some storage, something like firebase might work in that case.

Reply:
Yeah, but setting it up would be more time consuming than using something that already exists